### PR TITLE
Issue #1791: Join Row Layout

### PR DESCRIPTION
**Pull request by hawkfish**
Convert the join hash table to use `RowLayout` and the `RowOperations` used by the aggregate hash tables. This includes the `Match` operation, which is now shared by both hash tables.

### CODE REVIEW

File: src/common/row_operations/row_gather.cpp [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	auto data = FlatVector::GetData<T>(col);
+	auto &col_mask = FlatVector::Validity(col);
+
+	for (idx_t i = 0; i < count; i++) {
+		auto row_idx = row_sel.get_index(i);
+		auto row = ptrs[row_idx];
+		auto col_idx = col_sel.get_index(i);
+		data[col_idx] = Load<T>(row + col_offset);
+		ValidityBytes row_mask(row);
+		if (!row_mask.RowIsValid(row_mask.GetValidityEntry(entry_idx), idx_in_entry)) {
```
Comment (Mytherin) [inline src/common/row_operations/row_gather.cpp:33]: For the future: this check could be optimized away if we know the column does not contain null values (e.g. through statistics).

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks great to me.

One minor note for the future.

Have you had the chance of running some benchmarks to see how this influences join performance (e.g. TPC-H)?

File: src/common/row_operations/row_gather.cpp (hunk 1 above)
Comment (hawkfish) [reply src/common/row_operations/row_gather.cpp:33]: Yeah, I noticed that the one of the scatter code implementations was not checking whether the block was all valid, and the shared implementation does this, so there have been some wins.

**Comment by hawkfish** [comment]
> Have you had the chance of running some benchmarks to see how this influences join performance (e.g. TPC-H)?

No but I should. (Especially since I wrote a general purpose Python module to parse the output!) I was hoping that converting to offsets would save cache writes, but I haven't measured it.

**Comment by Mytherin** [comment]
> No but I should. (Especially since I wrote a general purpose Python module to parse the output!) I was hoping that converting to offsets would save cache writes, but I haven't measured it.

Unfortunately the output of the benchmark_runner has since been converted to TSV format!

**Review by Mytherin — APPROVED** [review]

**Pull request merged**